bigint/src/lib.rs

```diff
@@ -1349,6 +1356,47 @@ impl Integer for BigUint {
     }
 }
 
+fn high_bits_to_u64(v: &BigUint) -> u64 {
+    match v.data.len() {
+        0   => 0,
+        1   => v.data[0] as u64,
+        _   => {
+            let mut bits = v.bits();
+            let mut ret = 0u64;
+            let mut ret_bits = 0;
+
+            for d in v.data.iter().rev() {
+                let digit_bits = (bits + big_digit::BITS - 1) %
+                    big_digit::BITS + 1;
```

Comment:
That's kind of how I feel :)

I can go back to the subtract, I just want to get this patch in and get to
more interesting stuff...

On Thu, Jul 14, 2016 at 9:53 PM, Josh Stone <notifications@github.com>
wrote:

> In bigint/src/lib.rs
> <https://github.com/rust-num/num/pull/204#discussion_r70924708>:
>
> > @@ -1349,6 +1356,47 @@ impl Integer for BigUint {
> >      }
> >  }
> >
> > +fn high_bits_to_u64(v: &BigUint) -> u64 {
> > +    match v.data.len() {
> > +        0   => 0,
> > +        1   => v.data[0] as u64,
> > +        _   => {
> > +            let mut bits = v.bits();
> > +            let mut ret = 0u64;
> > +            let mut ret_bits = 0;
> > +
> > +            for d in v.data.iter().rev() {
> > +                let digit_bits = (bits + big_digit::BITS - 1) %
> > +                    big_digit::BITS + 1;
>
> Yes. On 32-bit platforms, a BigUint that grows more than 512MB is going
> to have a bad time anywhere that's using ::bits(). One that's just on the
> edge could create the possible-but-unlikely addition overflow here. Meh.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-num/num/pull/204/files/a1772f2a0d3e8fb7384a89322c93816b017fcc57#r70924708>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB5r7kjuyGdzI_1w4gfKcARYpU5KQZxfks5qVyBOgaJpZM4JIeee>
> .
>